File: src/main/java/com/github/sshaddicts/lucrecium/datasets/ImageDataSet.java
package com.github.sshaddicts.lucrecium.datasets;

import org.datavec.api.io.labels.ParentPathLabelGenerator;
import org.datavec.api.split.FileSplit;
import org.datavec.image.recordreader.ImageRecordReader;
import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.nd4j.linalg.dataset.api.preprocessor.DataNormalization;
import org.nd4j.linalg.dataset.api.preprocessor.ImagePreProcessingScaler;

import java.io.File;
import java.io.IOException;
import java.util.Random;

//TODO refactor
public class ImageDataSet {

    private DataSetIterator iterator;

    private final int outputLabelCount;
    private final int batchSize;

    private ImageRecordReader recordReader;

    public ImageRecordReader getRecordReader() {
        return recordReader;
    }

    public ImageDataSet(int outputLabelCount, int batchSize) {
        this.outputLabelCount = outputLabelCount;
        this.batchSize = batchSize;
    }

    public void initFromDirectory(String parentDir) throws IOException {
        File parentDirectory = new File(parentDir);

        String[] allowedExtensions = new String[]{".png", ".jpg"};
        FileSplit fileSplit = new FileSplit(parentDirectory, allowedExtensions, new Random());

        ParentPathLabelGenerator labelMarker = new ParentPathLabelGenerator();

        this.recordReader = new ImageRecordReader(32, 32, 1, labelMarker);

        this.iterator = new RecordReaderDataSetIterator(recordReader, batchSize, 1, outputLabelCount);
        DataNormalization scaler = new ImagePreProcessingScaler(0,1);
        scaler.fit(this.iterator);
        this.iterator.setPreProcessor(scaler);
        recordReader.initialize(fileSplit);
    }

    public DataSetIterator getIterator() {
        return iterator;
    }
}

File: src/main/java/com/github/sshaddicts/lucrecium/datasets/ImageDataSet.kt
package com.github.sshaddicts.lucrecium.datasets

import org.datavec.api.io.labels.ParentPathLabelGenerator
import org.datavec.api.split.FileSplit
import org.datavec.image.recordreader.ImageRecordReader
import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator
import org.nd4j.linalg.dataset.api.preprocessor.ImagePreProcessingScaler
import java.io.File
import java.io.IOException
import java.util.*

class ImageDataSet(private val outputLabelCount: Int, private val batchSize: Int) {

    var iterator: DataSetIterator? = null
        private set

    var recordReader: ImageRecordReader? = null
        private set

    @Throws(IOException::class)
    fun initFromDirectory(parentDir: String) {
        val parentDirectory = File(parentDir)

        val allowedExtensions = arrayOf(".png", ".jpg")
        val fileSplit = FileSplit(parentDirectory, allowedExtensions, Random())

        val labelMarker = ParentPathLabelGenerator()

        this.recordReader = ImageRecordReader(32, 32, 1, labelMarker)

        this.iterator = RecordReaderDataSetIterator(recordReader, batchSize, 1, outputLabelCount)
        val scaler = ImagePreProcessingScaler(0.0, 1.0)
        scaler.fit(this.iterator)
        this.iterator!!.preProcessor = scaler
        recordReader!!.initialize(fileSplit)
    }
}
